src/qibolab/backends.py

```diff
@@ -26,13 +27,33 @@ def __init__(self, platform, runcard=None):
             "numpy": self.np.__version__,
             "qibolab": qibolab_version,
         }
+        self.compiler = Compiler.default()
 
     def apply_gate(self, gate, state, nqubits):  # pragma: no cover
         raise_error(NotImplementedError, "Qibolab cannot apply gates directly.")
 
     def apply_gate_density_matrix(self, gate, state, nqubits):  # pragma: no cover
         raise_error(NotImplementedError, "Qibolab cannot apply gates directly.")
 
+    def assign_measurements(self, measurement_map, circuit_result):
+        """Assigning measurement outcomes to :class:`qibo.states.MeasurementResult` for each gate.
+
+        This allows properly obtaining the measured shots from the :class:`qibo.states.CircuitResult`
+        object returned by the circuit execution.
+
+        Args:
+            measurement_map (dict): Map from each measurement gate to the sequence of
+                readout pulses implementing it.
+            circuit_result (:class:`qibo.states.CircuitResult`): Circuit result object
+                containing the readout measurement shots. This is created in ``execute_circuit``.
+        """
+        readout = circuit_result.execution_result
+        for gate, sequence in measurement_map.items():
+            _samples = map(lambda pulse: readout[pulse.serial].shots, sequence.pulses)
+            samples = list(filter(lambda x: x is not None, _samples))
+            gate.result.backend = self
```

Comment:
Unfortunately this is a qibo thing that I cannot fix here. The reason the result objects need to have access to the backend is that when you do things like
```py
result.frequencies()
```
some calculation that invokes the backend may need to happen, so the backend needs to be accessible in the result.

Also, because the circuit and gate construction is backend agnostic, when you do
```py
c = Circuit(2)
m = c.add(gates.M(0))
```
`m: MeasurementResult` does not know in advance the backend that you will use to execute the circuit. So we need to assign the backend to `m` after we execute the circuit. This is what line 54 does.

Reply:
It is usually a good idea to keep the information flow in a single direction, to avoid loops, but if that's the way to do, I wouldn't stop the whole PR because of this.